Ensuring pytest catches pipeline errors

Would be nice to know what you think about this one @jsoriano .

With both pytest and nosetest there has been one small change in need that I hope would be easy to implement, and that is the possibility for it to catch "HTTP Code 400" returned by Elasticsearch when running test_modules or test_xpack_modules.

For each module that is tested, before ingesting the test files and comparing it with the golden files, it will always have to install the pipeline and index pattern, however, even when pipeline installation fails it will continue trying until it times out, especially in CI processes where timeout in some cases is increased this creates unnecessary wait times.

Would it be possible to implement this check? If you don't have time could you point me towards the part handling pipeline installations and I can create a PR around it?

Example output we want to catch early but don't:

```
cat build/system-tests/run/test_xpack_modules.XPackTest.test_fileset_file_2_azure/output.log | grep ERROR
2020-09-06T06:53:38.648-0200	ERROR	[publisher_pipeline_output]	pipeline/output.go:154	Failed to connect to backoff(elasticsearch(http://elasticsearch:9200)): Connection marked as failed because the onConnect callback failed: 1 error: Error loading pipeline for fileset azure/activitylogs: couldn't load pipeline: couldn't load json. Error: 400 Bad Request: {"error":{"root_cause":[{"type":"parse_exception","reason":"[type] required property is missing","property_name":"type","processor_type":"convert"}],"type":"parse_exception","reason":"[type] required property is missing","property_name":"type","processor_type":"convert"},"status":400}. Response body: {"error":{"root_cause":[{"type":"parse_exception","reason":"[type] required property is missing","property_name":"type","processor_type":"convert"}],"type":"parse_exception","reason":"[type] required property is missing","property_name":"type","processor_type":"convert"},"status":400}
2020-09-06T06:53:42.509-0200	ERROR	[publisher_pipeline_output]	pipeline/output.go:154	Failed to connect to backoff(elasticsearch(http://elasticsearch:9200)): Connection marked as failed because the onConnect callback failed: 1 error: Error loading pipeline for fileset azure/activitylogs: couldn't load pipeline: couldn't load json. Error: 400 Bad Request: {"error":{"root_cause":[{"type":"parse_exception","reason":"[type] required property is missing","property_name":"type","processor_type":"convert"}],"type":"parse_exception","reason":"[type] required property is missing","property_name":"type","processor_type":"convert"},"status":400}. Response body: {"error":{"root_cause":[{"type":"parse_exception","reason":"[type] required property is missing","property_name":"type","processor_type":"convert"}],"type":"parse_exception","reason":"[type] required property is missing","property_name":"type","processor_type":"convert"},"status":400}
```

The above error will not cause any test to fail, which is weird since without the pipeline everything else will 100% fail always. While it should not stop it from testing further on other modules, it should stop testing for that one module and fail it with a descriptive error message, currently the output from this is:

```
(2 durations < 0.005s hidden.  Use -vv to show these durations.)
==================================================================== short test summary info ====================================================================
FAILED tests/system/test_xpack_modules.py::XPackTest::test_fileset_file_2_azure - Failed: Timeout >90.0s
FAILED tests/system/test_xpack_modules.py::XPackTest::test_fileset_file_3_azure - beat.beat.TimeoutError: Timeout waiting for 'cond' to be true. Waited 10 sec..
```

--- a/filebeat/tests/system/test_modules.py
+++ b/filebeat/tests/system/test_modules.py
@@ -139,22 +139,29 @@ class Test(BaseTest):
             cmd.append("{module}.{fileset}.var.format=json".format(module=module, fileset=fileset))
 
         output_path = os.path.join(self.working_dir)
-        output = open(os.path.join(output_path, "output.log"), "ab")
-        output.write(bytes(" ".join(cmd) + "\n", "utf-8"))
+        # Runs inside a with block to ensure file is closed afterwards
+        with open(os.path.join(output_path, "output.log"), "ab") as output:
+            output.write(bytes(" ".join(cmd) + "\n", "utf-8"))
 
-        # Use a fixed timezone so results don't vary depending on the environment
-        # Don't use UTC to avoid hiding that non-UTC timezones are not being converted as needed,
-        # this can happen because UTC uses to be the default timezone in date parsers when no other
-        # timezone is specified.
-        local_env = os.environ.copy()
-        local_env["TZ"] = 'Etc/GMT+2'
+            # Use a fixed timezone so results don't vary depending on the environment
+            # Don't use UTC to avoid hiding that non-UTC timezones are not being converted as needed,
+            # this can happen because UTC uses to be the default timezone in date parsers when no other
+            # timezone is specified.
+            local_env = os.environ.copy()
+            local_env["TZ"] = 'Etc/GMT+2'
 
-        subprocess.Popen(cmd,
-                         env=local_env,
-                         stdin=None,
-                         stdout=output,
-                         stderr=subprocess.STDOUT,
-                         bufsize=0).wait()
+            subprocess.Popen(cmd,
+                             env=local_env,
+                             stdin=None,
+                             stdout=output,
+                             stderr=subprocess.STDOUT,
+                             bufsize=0).wait()
+
+        # List of errors to check in filebeat output logs
+        errors = ["Error loading pipeline for fileset"]
+        # Checks if the output of filebeat includes errors
+        contains_error, error_line = file_contains(os.path.join(output_path, "output.log"), errors)
+        assert contains_error is False, "Error found in log:{}".format(error_line)
 
         # Make sure index exists
         self.wait_until(lambda: self.es.indices.exists(self.index_name))
@@ -311,6 +318,15 @@ def clean_keys(obj):
 def delete_key(obj, key):
     if key in obj:
         del obj[key]
+
+
+def file_contains(filepath, strings):
+    with open(filepath, 'r') as file:
+        for line in file:
+            for string in strings:
+                if string in line:
+                    return True, line
+    return False, None
 
 
 def pretty_json(obj):

